poetry_to_pre_commit/sync_hooks_additional_dependencies.py

```diff
@@ -68,12 +74,37 @@ def get_poetry_deps(*, cwd: pathlib.Path | None = None, group: str) -> Iterable[
         yield f"{dep.complete_name}=={package.version}"
 
 
+def update_or_remove_additional_deps(
+    poetry_deps: set[str], hook_additional_deps: list[str]
+) -> set[str]:
+    # Additional packages that are already in pre-commit configuration could be listed with
+    # any format that is accepted by pip - use `Requirement` to parse them properly.
+    current_deps = [Requirement(dep).name for dep in hook_additional_deps]
+
+    return {
+        package
+        for package in poetry_deps
+        # package is yielded by `get_poetry_deps` above, and we are pretty sure that this won't raise `IndexError`
+        if package.split("==")[0].split("[")[0] in current_deps
```

Comment:
This `package.split` is quite ugly. I don't know what else to do without modifying `get_poetry_deps`, but if you are OK I would to make it yield `Dependency` object instead of strings. Basically, to replace:
```
        yield f"{dep.complete_name}=={package.version}"
```
with:
```
        yield dep
```
So that I can do `package.complete_name` here.

Reply:
I think the problem is that what we pass to this function does not fit anymore what we need.

We should recieve a set of [`PoetryPackages`](https://github.com/ewjoachim/poetry-to-pre-commit/blob/main/poetry_to_pre_commit/common.py#L27C7-L27C20) and that class should be updated to also have optional `extras` (`set[str]`). There should be a function (or a class of `PoetryPackage`) that turns a `PoetryPackage` into a requirement string (sort the extras by name to ensure consistency).
This way, we can filter the `PoetryPackages` using `current_deps` (which should be a set rather than a list).

Reply:
OK, I realize that my suggestion to `yield dep` would not work, because we then don't have the pinned version.
I started to refactor following your suggestions, but this is actually too much changes for this PR. This requires to adapt several things including the tests because `set` and `PoetryPackage` are not hashable etc. I would be OK to do this in another PR.

PS: have you considered to organize your hooks with classes? I find it easier to read than several functions, because it allows to scope them together + shorten the function names (since we already get the common info from the class). But I guess this is a matter of personal preference and don't want to debate that :-) It's just a suggestion. I know that you prefer to avoid classes in the test files, but I was wondering if you would like them in the actual code.

Reply:
Ok... So I'm a bit lot: is the PR in a state you'd consider mergeable ?

Reply:
Let's merge, then :)
The stakes are still quite low, so...